### PR TITLE
LOXI-69: Ignore OF Objects with unknown discriminator values

### PR DESCRIPTION
**Pull request by andi-bigswitch**
Reviewer: @Sovietaced @rlane 

This pull request enables us to ignore OF Messages that we do not understand (because we don't know the discriminator value). 

The previous behavior was to disconnect the switch if it sends a message we do not understand.

This PR solves the problem, but I really dislike how big and clunky and just ugly it is. I'm putting it up for discussion here. I'm not necessarily suggesting we merge it in the present form.

### CODE REVIEW

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com//job/loxigen_pull_req/178/
Test PASSed.

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/handler/UnparsedHandlers.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    public static void setDefaultHandler(UnparsedHandler handler) {
+        UnparsedHandlers.defaultHandler = handler;
+    }
+
+    public static UnparsedHandler getDefaultHandler() {
+        return defaultHandler;
+    }
+
+
+    /** UnparsedHandler that keeps count of the unknown messages by class and discrimantor.
```
Comment (Sovietaced) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/handler/UnparsedHandlers.java]: super duper nit: discriminator*

File: java_gen/pre-written/src/test/java/org/projectfloodlight/openflow/protocol/ver13/OFMessageUnknownTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            LoggerFactory.getLogger(OFMessageUnknownTest.class);
+
+    final static byte[] UNKNOWN_MESSAGE_SERIALIZED =
+            new byte[] { 0x4, 077, 0x0, 0xb, 0x12, 0x34, 0x56, 0x78, 0x61, 0x62, 0x63 };
+
+    final static byte[] OXM_UNKNOWN_SERIALIZED =
+            new byte[] { (byte) 0x80, 0x0, /** tcp flags, introduced in OF1.5 */ 42 << 1, 0x2, 0x0, 0x0};
+
+    @Test
+    public void testSkipUnknownMessage() throws OFParseError {
```
Comment (Sovietaced) [inline java_gen/pre-written/src/test/java/org/projectfloodlight/openflow/protocol/ver13/OFMessageUnknownTest.java]: Should add tests for the other UnparsedHandlers

File: java_gen/templates/of_virtual_class.java [hunk 3]
```diff
@@ -92,16 +100,40 @@
...
                case ${m.priv_value}:
+                   bb.readerIndex(start);
                    // discriminator value ${m.enum_value}=${m.value} for class ${sub.name}
-                   return ${sub.name}.READER.readFrom(bb);
+                   return ${sub.name}.READER.readFrom(context, bb);
 //:: #endif    # generate_class
 //:: #endfor
                default:
-                   throw new OFParseError("Unknown value for discriminator ${prop.name} of class ${msg.name}: " + ${prop.name});
+                   context.getUnparsedHandler().unparsedMessage(${msg.name}.class, "${prop.name}", ${prop.name});
```
Comment (Sovietaced) [inline java_gen/templates/of_virtual_class.java]: So this is where the calls will start.. looks good.

**Comment by andi-bigswitch** [comment]
Not working on this right now.

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/145/

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-425-jenkins-loxigen_pull_req_artifacts-128

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-425-jenkins-loxigen_pull_req_artifacts-129

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/146/

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_artifacts/129/

**Comment by Sovietaced** [comment]
@andi-bigswitch Let me know when this is ready for review.

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b1

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b2

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b3

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b4

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b5

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b6

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b7

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b8

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b9

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b10

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b11

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b12

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b13

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b14

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b15

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b16

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b17

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b18

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b19

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b20

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-425-b21

**Pull request closed without merging**